### DISCUSSION TITLE
invalitate/clear whole api data on user logout

### DISCUSSION THREAD

**Discussion by wuifdesign** [reduxjs/redux-toolkit#1406] [Q&A]
Hi,

I want to invalidate a whole api if a user gets logged out (cookie expired and the server response 401).
I extended the baseQuery to do this:

```
const baseQuery = fetchBaseQuery({ baseUrl: '/api' })
const baseQueryWithAuthCheck: BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError> = async (
  args,
  api,
  extraOptions
) => {
  const result = await baseQuery(args, api, extraOptions)

  if (result.error && result.error.status === 401) {
    // should invalitate whole api state
    // goto auth page
    retry.fail(result.error)
  }

  return result
}
```

First way would be to use tags, but i can't find a way to `invalidatesTags` from outside a mutation and invalidation will still keep the data in the redux store (if i understand it right). i would need something like `api.invalidateTags(['Auth'])` and add the `Auth` tag to all endpoints. but still i will not be able to invalitate it inside the `baseQueryWithAuthCheck`.

Better way would be to get the data out of the redux store. something like `api.clear()` which removes all stored data and triggers refetch for all active subscriptions.

**Comment by phryneas** [comment]
Are you looking for https://redux-toolkit.js.org/rtk-query/api/created-api/cache-management-utils#resetapistate ?

**Reply by wuifdesign** [reply]
ahh. thanks. i used a whole hour to search the api docs/issues/discussions but didn't find it. maybe i just read over it. XD 

will this trigger automatic refetch?

**Reply by phryneas** [reply]
Good question, I've only used it in test environments for resetting until now. You might want to give it a try and report back.

An alternative would be having an `AUTHENTICATED` tag type on all authenticated apis and invalidating that on logout.

**Reply by wuifdesign** [reply]
`resetApiState` does not trigger refetch. so it's not the way to go. would be nice to have something like, but this triggering refetch. (or maybe somthing like react-query https://react-query.tanstack.com/guides/query-invalidation where you invalidate all data if you dont pass any parameters to `invalidateTags`)

Also i have a problem with writing an `invalidateTags` dispatch. as the api isn't defined at this point, and the `api` passed to the `baseQuery` doesn't include the utils. It only includes `{signal: ..., dispatch: ..., getState: ...}` 

```
const baseQuery = fetchBaseQuery({ baseUrl: '/api' })
const baseQueryWithAuthCheck: BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError> = async (
  args,
  api,
  extraOptions
) => {
  const result = await baseQuery(args, api, extraOptions)

  if (result.error && result.error.status === 401) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    api.dispatch(backendApi.util.invalidateTags(['AUTHENTICATED'])) // backendApi isn't defined here. it will work, but not a clean code
    retry.fail(result.error)
  }

  return result
}

export const backendApi = createApi({
  reducerPath: 'backendApi',
  baseQuery: baseQueryWithAuthCheck,
  tagTypes: ['User'],
  endpoints: (builder) => ({
    ...
  })
})
```

**Reply by phryneas** [reply]
What about
```js
let backendApi;
const baseQueryWithAuthCheck = ...
backendApi = createApi({
```
?

The baseQuery will never be invoked during `createApi`

**Reply by wuifdesign** [reply]
Than i have to do the typing of backendApi all by myself. otherwise i will get: `Variable 'backendApi' implicitly has an 'any' type.`.
Loosing all autocompletion. And doing the types by myself for all endpoints is really hard.

```
let backendApi: Api<
  BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError, {}, {}>,
  {
    getUser: QueryDefinition<
      void,
      BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError, {}, {}>,
      'AUTHENTICATED' | 'User',
      UserType,
      string
    >
  },
  'backendApi',
  'AUTHENTICATED' | 'User',
  typeof coreModuleName | typeof reactHooksModuleName
>
```

this will explode if i add more endpoints.

**Reply by phryneas** [reply]
hmm. like this?
```ts
let referenceToApi: undefined | typeof backendApi;
const baseQueryWithAuthCheck = ...
const backendApi = createApi({...});
referenceToApi = backendApi
```

**Reply by wuifdesign** [reply]
if i do this i will get `ESLint: 'referenceToApi' is never reassigned. Use 'const' instead.(prefer-const)` as it is used before it is assigned (inside the baseQueryWithAuthCheck). even is i assign it afterwards.

but i may just supress it with `// eslint-disable-next-line prefer-const`.

**Reply by phryneas** [reply]
Yeah, can't do anything about linter rules unfortunately - I think it's okay to just disable that there :)
